lib/rules/max-statements-per-line.js

```diff
@@ -0,0 +1,119 @@
+/**
+ * @fileoverview Enforce a maximum number of statements per line.
+ * @author Kenneth Williams
+ * @copyright 2016 Kenneth Williams. All rights reserved.
+ * @copyright 2016 Michael Ficarra. All rights reserved.
+ * See LICENSE file in root directory for full license.
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = function(context) {
+
+    var option = context.options[0],
+        maxStatementsPerLine = 1;
+
+    if (typeof option === "object" && option.hasOwnProperty("maximum") && typeof option.maximum === "number") {
+        maxStatementsPerLine = option.maximum;
+    }
+    if (typeof option === "number") {
+        maxStatementsPerLine = option;
+    }
+
+    //--------------------------------------------------------------------------
+    // Helpers
+    //--------------------------------------------------------------------------
+
+    /**
+     * Checks if two statements are on the same line
+     * @param {ASTNode} node1 The first node to check
+     * @param {ASTNode} node2 The second node to check
+     * @returns {boolean} true if both nodes exist and are on the same line
+     * @private
+     */
+    function isStatementsOnSameLine(node1, node2) {
+        if (!node1 || !node2) {
+            return false;
+        }
+        return node1.loc.end.line === node2.loc.start.line;
+    }
+
+    /**
+     * Reports a node
+     * @param {ASTNode} node The node to report
+     * @param {int} count Number of statements on line
+     * @param {int} max Maximum number of statements allowed
+     * @returns {void}
+     * @private
+     */
+    function report(node, count, max) {
+        context.report(
+            node,
+            "This line has too many statements ({{count}}). Maximum allowed is {{max}}.",
+            { count: count, max: max });
+    }
+
+    /**
+     * Enforce a maximum number of statements per line
+     * @param {ASTNode} node node to evaluate
+     * @returns {void}
+     * @private
+     */
+    function enforceMaxStatementsPerLine(node) {
+        var body = node.body;
+        if (body.length < 2) {
+            return;
+        }
+
+        var numberOfStatementsOnThisLine = 1;
+        var lastStatementLine = body[0].loc.end.line;
+        for (var i = 1, l = body.length; i < l; ++i) {
+            var currentStatement = body[i];
+            var nextStatement = body[i + 1];
+            if (currentStatement.loc.start.line <= lastStatementLine) {
+                ++numberOfStatementsOnThisLine;
+            } else {
+                numberOfStatementsOnThisLine = 1;
+                lastStatementLine = currentStatement.loc.end.line;
+            }
+            if (!isStatementsOnSameLine(currentStatement, nextStatement) &&
+                    numberOfStatementsOnThisLine > maxStatementsPerLine) {
```

Comment:
I'm not sure the user cares how many statements are on the line; just that the rule has been violated.